### ISSUE TITLE
Race condition between blur and focus causes active state to be lost

### ISSUE THREAD

**Issue by kpdecker** [erikras/redux-form#1001]
In my app I have a meta field that consists of multiple other fields (in this case, checkboxes). Rather than sending many focus and blur operations, I try to filter this to a single external `onFocus` and `onBlur` call via a setTimeout, i.e:

```
  onFocus() {
    if (!this.props.onFocus) {
      return;
    }

    clearTimeout(this.blurTimeout);

    if (!this.focused) {
      this.focused = true;
      this.props.onFocus(this.props.value);
    }
  }
  onBlur() {
    if (!this.props.onBlur) {
      return;
    }

    this.blurTimeout = setTimeout(() => {
      this.focused = false;
      this.props.onBlur(this.props.value);
    }, 50);
  }
```

This pattern apparently exposes a race condition between `onBlur` and `onFocus` calls between two fields. The timeout above causes `onBlur` to occur after `onFocus` for the field that receives focus, causing the `_active` state to be removed even though the other field still maintains focus at the user level. So my question is if the blur handler should check to ensure that it's not interacting with the form-level state when it's no longer the target field?

Something like:

```
  [BLUR](state, {field, value, touch}) {
    // remove _active from state
    const {_active, ...stateCopy} = state;  // eslint-disable-line prefer-const
    if (field !== _active) {
      stateCopy._active = _active;
    }
```

**Comment by erikras** [comment]
In my experience, the `BLUR` is _always_ dispatched before the `FOCUS`. Obviously, your `setTimeout` call messes with that.

>  I try to filter this to a single external `onFocus` and `onBlur` call via a `setTimeout`

Why? This sounds like [probably-unwarranted fear of dispatching many actions](https://github.com/erikras/redux-form/issues/991#issuecomment-219774646). Can you explain why you think this is a good idea?

**Comment by kpdecker** [comment]
This is to avoid marking the fields as touched until they are "fully" blurred. Basically I don't want some of the validation logic that is place to trigger until the user is focused on something completely different.

**Comment by ooflorent** [comment]
I had experienced this race condition in an application. The issue is not that `blur` is dispatched after `focus` but the way the reducer handles the active field.

```js
// This is not a working snippet.
// It illustrates what's going on here.

// Action log
focus("form", "foo")
focus("form", "bar")
blur("form", "foo")

// Result state
{
  foo: {_active: false},
  bar: {_active: false}, // expected `true`
}
```

**Comment by ooflorent** [comment]
I've created to test case to reproduce the issue. I'll try to submit a patch.

**Comment by ooflorent** [comment]
@erikras Is it worth fixing this? I'm asking because there are 2 issues here:

1. A race condition where `focus` is dispatched before `blur`.
    We can't do anything for this. It is a React behavior. We cannot predict or prevent this.

2. The `reducer` clears `_active` even if the blurred field is not the active one.
    This is a long time behavior. Is it what we expect? If I change this it may break existing apps since the _bug_ is here for ages!

**Comment by kpdecker** [comment]
I hadn't imagined that a fix for this would go into anything prior to 6.x, so (2) legacy behavior is less of a concern, no?

If (2) is not fixed, `active` has known cases that could break down in surprising manners for developers, it certainly caused myself a fair amount of grief, although perhaps there are other ways to solve my use case in a different manner. Regardless, the library should do what it can to protect users from themselves, rather than hoping that they don't try to be creative like I did :)

**Comment by ooflorent** [comment]
@kpdecker I encountered this issue without being creative and fixed it in a hacky way (using `setTimeout`).

**Comment by erikras** [comment]
Released as [`v5.2.4`](https://github.com/erikras/redux-form/releases/tag/v5.2.4).
